File: 프로그래머스/lv1/42889. 실패율/실패율.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

class Solution {
    public int[] solution(int N, int[] stages) {
        int[] answer = new int[N];
        
        int players = stages.length;
        int[] numberOfPlayersOnStage = new int[N+2];
        int[] numberOfPlayersCleared = new int[N+2];
        for(int i : stages) numberOfPlayersOnStage[i]++;
        for(int i = 1; i < N+2; i++) 
            for(int j = i; j < N+2; j++)
                numberOfPlayersCleared[i] += numberOfPlayersOnStage[j];
        
        double[] failureRate = new double[N+2];
        for(int i = 1; i < N+2; i++){
            if(numberOfPlayersCleared[i] == 0) {
                failureRate[i] = 0;
                continue;
            }
            failureRate[i] = (double)numberOfPlayersOnStage[i] / numberOfPlayersCleared[i];
        }
        
        ArrayList<stageFailureRate> al = new ArrayList<>();
        for(int i = 1; i < N+1; i++) 
            al.add(new stageFailureRate(i, failureRate[i]));
        
        Collections.sort(al, new Comparator<stageFailureRate>(){ 		
			public int compare(stageFailureRate sf1, stageFailureRate sf2) { 			
				if(sf1.failureRate == sf2.failureRate)			
					return Integer.compare(sf1.stage, sf2.stage);
				return Double.compare(sf2.failureRate, sf1.failureRate);
			}
		});
        
        int idx = 0;
        for(stageFailureRate sf : al) answer[idx++] = sf.stage;
        return answer;
    }
    
    class stageFailureRate {
        int stage;
        double failureRate;
        stageFailureRate(int stage, double failureRate) {
            this.stage = stage;
            this.failureRate = failureRate;
        }
    }
}

File: 프로그래머스/1/42889. 실패율/실패율.py
def solution(N, stages):
    answer = []
    
#     failureRate = [[0, 0, idx] for idx in range(N + 2)]    # not clear yet, arrived
#     for s in stages:
#         failureRate[s][0] += 1
#         for i in range(1, s + 1): failureRate[i][1] += 1
    
#     for idx in range(N + 2): 
#         b = failureRate[idx][1]
#         if b == 0: failureRate[idx][1] = 1
    
#     failureRate = failureRate[1:-1]
#     failureRate.sort(key = lambda x: -x[0]/x[1])
    
#     for a, b, idx in failureRate: answer.append(idx)

    fails = {}
    total = len(stages)
    for stage in range(1, N + 1):
        if total != 0:
            cnt = stages.count(stage)
            fails[stage] = cnt / total
            total -= cnt
        else:
            fails[stage] = 0
    
    answer = sorted(fails, key=lambda x: fails[x], reverse=True)
    
    return answer